docs: Explain new primary selection modes (#241)
akka/akka-sdk: docs/src/modules/concepts/pages/multi-region.adoc

* Low latency global reads
* Low latency global writes

Akka has two replication modes; replicated reads and replicated writes.

== Replicated reads

Akka's replicated reads offers full data replication across regions and even cloud providers, without any changes to the service implementation: an entity has its "home" in one _primary region_, while being replicated to multiple other regions.

Read requests are always handled locally within the region that they occur.

An entity can only be written within a single region, known as the primary region.

> How Akka assigns the primary region to an entity is configurable. There are two modes for primary selection: **pinned-region-writes** and **request-region-writes**.

Author: We must explain the primary selection modes first, before presenting the example. And the current example is only valid for pinned writes.

+
  This is useful for scenarios where you want one primary region, with the ability to fail over to another region in the case of a regional outage.

Request-region primary selection mode::
  The primary region changes when another region receives a write request. Upon a write request to an entity in a region that is not the primary it will move its primary. The new primary ensures that all preceding events from the previous primary have been fully replicated and applied (i.e. persisted) before writing the new event, and thereby guarantees strong consistency when switching from one region to another. Subsequent write requests to the primary region of the entity are handled locally without any further coordination. Write requests to other regions will trigger the same switch-over process. All other entity instances operate unimpeded during the switch-over process.
+

>   This is useful for scenarios where you want to have the primary region for you data close to the users who use the data. A user, Alice, in the USA, will have her data in the USA, while a user Bob, in the UK, will have his data, in the UK. If Alice travels to Asia the data will follow her.

Reviewer: ```suggestion
  This is useful for scenarios where you want to have the primary region for your data close to the users who use the data. A user, Alice, in the USA, will have her data in the USA, while a user Bob, in the UK, will have his data, in the UK. If Alice travels to Asia the data will follow her.
```

Meanwhile, all requests made by Alice on her data are handled locally image:steps-1.svg[width=20] and get an immediate reply image:steps-2.svg[width=20]. The write operations are being replicated to the UK image:steps-3.svg[width=20].

== Replicated writes

The replicated write replication strategy allows every region to be capable of handling writes for all entities. This is done through the use of CRDTs, which can be modified concurrently in different regions, and their changes safely merged without conflict.

> == Replication Guarantees

Author: this is a new section that explains how it works, mostly to build trust that the mechanism is sane and solid

Meanwhile, all requests made by Alice on her data are handled locally image:steps-1.svg[width=20] and get an immediate reply image:steps-2.svg[width=20]. The write operations are being replicated to the UK image:steps-3.svg[width=20].

== Replicated writes

The replicated write replication strategy allows every region to be capable of handling writes for all entities. This is done through the use of CRDTs, which can be modified concurrently in different regions, and their changes safely merged without conflict.

== Replication Guarantees

Akka guarantees that all events created within one region are eventually replicated to all other regions in the project.

Each entity’s state is a series of events that are persisted in a local event journal, which acts as the source of events that must be replicated from one region to another. Having a durable, local event journal is the foundation for how Akka can recover an entity’s state in the event of failure.

> Each event has a sequence number that is validated on the receiving side to guarantee correct ordering and exactly-once processing of the events. A replicated event is processed by entities in other regions by having it added to its local event journal. Once added to the local event journal, it can be used to handle read requests in those regions.

Reviewer: The ambiguous `it`s make this paragraph a little more difficult to read. And maybe clarify the last `it` to being not just the event, but the derived state. Suggestion:

```suggestion
Each event has a sequence number that is validated on the receiving side to guarantee correct ordering and exactly-once processing of the events. A replicated event is processed by entities in other regions by having the event added to the local event journal of each entity. Once added to the local event journal, the replicated event can be used to update the entity's state and handle read requests in those regions.
```
